server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -147,149 +147,202 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
-        final long startTime = relativeTime();
-        final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
-
-        boolean hasIndexRequestsWithPipelines = false;
-        final MetaData metaData = clusterService.state().getMetaData();
-        ImmutableOpenMap<String, IndexMetaData> indicesMetaData = metaData.indices();
-        for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
-            IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
-            if (indexRequest != null) {
-                // get pipeline from request
-                String pipeline = indexRequest.getPipeline();
-                if (pipeline == null) {
-                    // start to look for default pipeline via settings found in the index meta data
-                    IndexMetaData indexMetaData = indicesMetaData.get(actionRequest.index());
-                    // check the alias for the index request (this is how normal index requests are modeled)
-                    if (indexMetaData == null && indexRequest.index() != null) {
-                        AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
-                        if (indexOrAlias != null && indexOrAlias.isAlias()) {
-                            AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
-                            indexMetaData = alias.getWriteIndex();
-                        }
+        new BulkExecutor(task, bulkRequest, listener).run();
+    }
+
+    /**
+     * A runnable that will ensure the cluster state has been recovered enough to
+     * read index metadata and templates/pipelines.  Will retry up to the bulk's timeout
+     */
+    private final class BulkExecutor extends ActionRunnable<BulkResponse> {
+        private final ClusterStateObserver recoveredObserver;
+        private final BulkRequest bulkRequest;
+        private final Task task;
+        long startTime;
+
+        BulkExecutor(Task task, BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+            super(listener);
+            this.recoveredObserver = new ClusterStateObserver(clusterService, bulkRequest.timeout(), logger, threadPool.getThreadContext());
+            this.bulkRequest = bulkRequest;
+            this.task = task;
+            startTime = relativeTime();
+        }
+
+        @Override
+        protected void doRun() {
+            ClusterState currentState = clusterService.state();
+            ClusterBlockException blockException = currentState.blocks().globalBlockedException(ClusterBlockLevel.WRITE);
+
+            if (blockException != null) {
+                recoveredObserver.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState newState) {
+                        // predicate passed, begin preparing for the bulk
+                        prepForBulk(newState);
```

Comment:
This is now sometimes on the cluster applier thread, which is probably not what we want. We should add this to the top of `prepForBulk`, and then make sure to call `prepForBulk` on the `WRITE` threadpool here.

```
assert ClusterApplierService.assertNotClusterStateUpdateThread("TransportBulkAction#prepForBulk");
```